feat(discover): Add auto aggregations

wmak opened the pull request:
- This will have discover/performance automatically add aggregations if
  they're used in a query
  - We currently pass a large number of selected columns for the trends
    queries just in case they get used, this way we only add those when
    they're used instead
  - eg. absolute_correlation is no longer added automatically
- This removes the extra fields from the trends endpoint calls
- This adds aggregations with no parameters or default parameters functions to trend autocomplete 
  - Also not including trendView in the fields anymore since its including the functions
  from the other performance views, can add these back if we want
  though.

ex.
absolute_correlation is in the auto complete now
<img width="1510" alt="image" src="https://user-images.githubusercontent.com/4205004/95927484-d3819700-0d8c-11eb-8cb9-024eb238765a.png">

### before 
<img width="724" alt="image" src="https://user-images.githubusercontent.com/4205004/95927618-43901d00-0d8d-11eb-9b4b-178032a4058f.png">

### after
~80% previous size
<img width="727" alt="image" src="https://user-images.githubusercontent.com/4205004/95927640-530f6600-0d8d-11eb-9a03-bf480ce7f710.png">

wmak committed:
feat(trends): Add auto aggregations

- This will have trends automatically add a set list of aggregations if
  they're used in a query
  - We currently pass a large number of selected columns for the trends
    queries just in case they get used, this way we only add those when
    they're used instead
  - eg. here already absolute_correlation is no longer added
    automatically

wmak committed:
ref(trends): Removing the extra fields from the trends endpoint

wmak committed:
feat(trends): Add correct functions to trend autocomplete

- Also not including trendView here since its including the functions
  from the other performance views, can add these back if we want
  though.

wmak committed:
fix: Checking having exists before using it

wmak commented on src/sentry/api/event_search.py:
```diff
@@ -2015,13 +2015,18 @@ def resolve_field(field, params=None):
+    Auto aggregates are aggregates that will be automatically added to the
+    list of aggregations when they're used in a condition. This is so that
+    they can be used in a condition without having to manually add the
+    aggregate to a field.
```
As discussed in slack, I'll enable this on both for all aggregates when there's at least one aggregate so that we aren't changing grouping on users in unexpected ways.
:+1:

wmak committed:
feat(discover): Expand auto aggregations to any aggregation

- This will now add any aggregations instead of from a set list
- Auto aggregations are only added when there's at least one other
  aggregation, this is so that we don't changing the result groupings,
  but generally for the performance views should be helpful.

wmak committed:
ref: Adding aggregates to autocomplete when available

markstory commented on src/sentry/eventstore/base.py:
```diff
@@ -27,6 +27,8 @@ class Filter(object):
     aggregations (Sequence[Any, str|None, str]): Aggregate functions to fetch.
     groupby (Sequence[str]): List of columns to group results by
+
+    _condition_aggregates (Sequence[str]): List of aggregates used in the condition
```
Why the `_`?

wmak replied:
🙃 woops accidental, removed in 44748fb

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -326,6 +327,9 @@ def query(
     auto_fields (bool) Set to true to have project + eventid fields automatically added.
+    auto_aggregations (bool) Whether aggregates should be added automatically if they're used
+                    in conditions, and there's at least one aggregate already.
+    use_aggregate_conditions (bool) Set to true if aggregates conditions should be used at all.
```
Should we error on `auto_aggregations=True and use_aggregate_conditions=False`?

wmak committed:
ref: Addressing PR comments

- Removing extra _ infront of condition aggregate comment
- Adding an assert that if we aren't using aggregate conditions, we
  don't try to use auto_aggregations

wmak committed:
ref: Only showing automatically added on auto_aggregation

wmak committed:
fix: Removing extra absolute_correlations

markstory commented on src/sentry/snuba/discover.py:
```diff
@@ -346,6 +346,9 @@ def query(
         if not use_aggregate_conditions:
+            assert (
+                not auto_aggregations
+            ), "Auto aggregations cannot be used without enabling aggregaet conditions"
```
```suggestion
            ), "Auto aggregations cannot be used without enabling aggregate conditions"
```

wmak committed:
Update src/sentry/snuba/discover.py

Co-authored-by: Mark Story <mark@sentry.io>
